### DISCUSSION TITLE
Best way to conditionally polyfill fetch

### DISCUSSION THREAD

**Discussion by goldingdamien** [node-fetch/node-fetch#1751] [Q&A] [answered]
Node.js had natively added fetch, but older versions don't accept it by default.
To make this transition without worry about the Node version, I would like a very basic script I can use to polyfill fetch if it is not available.

node-fetch provides this in the main README:
https://github.com/node-fetch/node-fetch#providing-global-access

- Is this the recommended way to achieve this?
- Does node-fetch automatically handle the conditional loading of fetch? Will node-fetch be used even if fetch already exists?

Regards.

**Comment by jimmywarting** [comment, marked as answer]
My personal suggestion would just be to tell folks to upgrade their NodeJS version.
Adding this to your package.json
```json
"engine": {
  "node": ">=18"
}
```
The alternative is that you let the clients bring in their own fetch impl from wherever they choose to bring in. weather it be undici, node-fetch v3 (or v2 for cjs) so you could let them decide for themself what they want to bring in. they might already have a dependency on undici, so it would be unnecessary to add node-fetch into the mix as well.

> Does node-fetch automatically handle the conditional loading of fetch? Will node-fetch be used even if fetch already exists?

No, but I have tough about doing that. it would certainly be a breaking change cuz they don't work _exactly_ the same, like passing a Agent as an option for example. 

if you want to support NodeJS v16 (which is the latest LTS then i would have opt'ed for using undici instead that's more spec compatible and uses web streams instead of node streams.
```js
if (!globalThis.fetch) {
  const { fetch, FormData, Headers, Request, Response } = await import('undici')
  Object.assign(globalThis, { fetch, FormData, Headers, Request, Response })
}
```
But i would also have added undici to optional dependencies so folks with newer NodeJS version don't need to download it in vain
undici is also what NodeJS have in their NodeJS core to bring fetch into the ecosystem

---

One thing if you are working with Blob/Files and wish to send multiple large files using FormData is to use `fs.openAsBlob(path, { type })` but that's only avaliable in NodeJS v20+
node-fetch export something similar to it (aka `blobFrom(path, { type })` with some variation of it, which works with native FormData but this can be imported independently from `fetch-blob` instead.

**Comment by goldingdamien** [comment]
Thank you for the detailed response.

Upgrading Node.js would be ideal, but even on a recent project there was a compatibility issue that meant I end up using Node 17, so it may be a while longer until that method can be relied upon.

I will check out undici. The code you provided seems like a good alternative polyfill.
Seems like the obvious answer if it is being used for fetch in Node.js.

Regards.
